### PR TITLE
Classifier themes

### LINKED ISSUE #4447: Text shown when hovering over back button is low contrast
## Expected behavior
Text should be high contrast with its background and easy to read.

## Current behavior
_Please include any error messages from the browser console and/or screenshots_
The text for this box is black on a grey background.

screen shot 2018-03-13 at 10 37 30 am

## Steps to replicate
1) go to a project with more than one task
2) on the second task hover over the back button

## Additional information

- **Operating system:**
- **Browser:**

### PR DESCRIPTION
**Pull request by srallen**
Staging branch URL: https://classifier-light-theme.pfe-preview.zooniverse.org/

This is the first half of the 'facelift' style update to the task area of the classifier. It should be noted that this does not entirely update the survey task according to the new design. 

This will close #4447.

TODO: 
- [x] Tests for new components
- [x] Dark theme. I'm branching off of this and doing that separately because this PR is getting big on its own. This could wait to merge until the dark theme is done and merged into this OR we can merge them separately.
- [x] Manually test in various browsers

Questions for @beckyrother:
1. What should be done with the restart mini-course button as seen in this test project?: https://classifier-light-theme.pfe-preview.zooniverse.org/projects/srallen086/workflow-assignment-testing/classify
2. What should the tutorial tab hover/focus state be? Right now it just opens the modal pop up like the old button
3. I guessed at a hover/focus state for the done/next buttons and made it a slightly lighter yellow. Is that ok?
4. I guessed on styling for the back button in the task navigation. You can see the back button on multi-step tasks starting on the second task. You can see it using the dev classifier: https://classifier-light-theme.pfe-preview.zooniverse.org/dev/classifier
5. What would you like done with the shortcut? I removed the button styling from it as it was styled like the darker style buttons from before and now just looks like a standard checkbox. It can be seen in the dev classifier, but in context it is most often used with survey tasks, so here's an example project: https://classifier-light-theme.pfe-preview.zooniverse.org/projects/srallen086/survey-test/classify
6. The focus/hover state for the question task looks like it has a gradient and/or opacity, but the inspector in invision isn't picking that up (seen here: https://projects.invisionapp.com/d/main#/console/12923997/270495244/preview) Can you confirm what you want done here?
7. Similarly the selected survey options appear to either have opacity or gradient, but invision inspector isn't picking that up either: https://projects.invisionapp.com/d/main#/console/12923997/270495286/preview

# Required Manual Testing

- [ ] Does the non-logged in home page render correctly?
- [ ] Does the logged in home page render correctly?
- [ ] Does the projects page render correctly?
- [ ] Can you load project home pages?
- [ ] Can you load the classification page?
- [ ] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [ ] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

### CODE REVIEW

**Comment by srallen** [comment]
I forgot to add, this also fixes #4460 by putting the selected theme into local storage. I'm not totally convinced that this setting should be complicated by putting it into user project preferences and/or have a project default, but I'm open to being convinced of that being useful or necessary.

**Comment by beckyrother** [comment]
@srallen This looks awesome, thank you! 

some answers:
1. Please use the text link style, same as “NEED SOME HELP WITH THIS TASK?”
2. Would it look weird with a teal focus state like the selected answer buttons?
3. Sorry, it should be #F6D885
4. That's perfect
5. I'm ok with the standard checkbox styling
6+7. Question hover (light teal): #8BCED2 to #C0E3E5; Next hover (gold): #F4CF00 to #E4960C; Secondary (dark teal): #00959C to #006F77. Here's all the buttons: https://invis.io/a/9S1A9Z112R6TYV
8. Rather than a teal box behind the text, just use Karla in the same style as the question, but #00979D.  

and a few questions:
1. Brooke brought up that some projects (like GZ) use images on their buttons and may want an option where the questions are left-aligned rather than centered. Is it possible to give project owners that option? Or say if images are used on the buttons, text is always left-aligned?

2. Looks like there's a shadow on the tutorial button, can that be removed? 
workflow-assignment-testing_ _classify_ _zooniverse

**Comment by eatyourgreens** [comment]
Using tab to navigate through the answers to a single question task, there's no visual focus indicator when I first tab to the answers. Once I select an answer with space, or up/down, the selected answer changes colour and I can see where focus is. Looks like there's no visual styling for the case where a hidden control has focus, but isn't selected.

Similarly, there's no focus indicator to tell me which answer I'm interacting with when I tab through the checkboxes for a multiple choice question. The visual style only changes when I select an answer. There should be a style for hover/focus.

No problems using the keyboard to navigate a survey task, testing with Snapshot Grumeti.

**Comment by beckyrother** [comment]
Thanks @eatyourgreens. On the projects I looked it, it seems like the default browser focus indicator is being used. Is this not the case? Should it be more custom? @srallen ?

**Comment by srallen** [comment]
@beckyrother I can demo what he means if you want to see it. The task area initially gets a focus outline, but tabbing into the actual answers don't have a visual focus because the radio button is being hidden. Browsers don't support CSS styling of the actual radio circle or checkbox square unfortunately, so we have to hide them somehow without breaking accessibility. I'm working on a fix at the moment. 

@eatyourgreens I borrowed the CSS to hide the radio button from the drawing task as well as the multi-image subject pip buttons to switch between frames. These might have the same issue.

**Comment by eatyourgreens** [comment]
@srallen @mcbouslog had a test organisation on staging, which had category radio buttons that used this technique too. I think it might have used slightly different CSS from the drawing tools.

**Comment by srallen** [comment]
Hm, I think I understand the difference between the subject pips, drawing tool selection, and the organization filter radio buttons. Those all start with one option already selected. For the single choice or multiple choice questions, they do not. So for the radio button groups that already have a default choice, it isn't as apparently that there isn't a keyboard focus state.

**Comment by beckyrother** [comment]
Is there a way to force a focus state on the button itself? Something like a 2px #005D69 border? 

classifier

**Comment by srallen** [comment]
@eatyourgreens @beckyrother I've updated the staged branch with handlers for the single choice and multiple choice tasks to have a proper keyboard focus state. This required a bit of editing of those tasks and I wrote a bunch of new tests to test the event handlers for both of them (onChange, onFocus, and onBlur). Becky, I'm still a bit confused about the styles I should be using 😬. I reused what I thought should be the hover/focus state for the task answer buttons, which is the teal gradient that you clarified a few comments above. Perhaps having a chat in person would help?

Here's the dev classifier: https://classifier-light-theme.pfe-preview.zooniverse.org/dev/classifier

Then a few projects to check this out:
The restart the minicourse button styled as requested: https://classifier-light-theme.pfe-preview.zooniverse.org/projects/srallen086/workflow-assignment-testing/classify
A single choice task: https://classifier-light-theme.pfe-preview.zooniverse.org/projects/srallen086/galaxy-zoo-in-astronomy-101/classify
A drawing task: https://classifier-light-theme.pfe-preview.zooniverse.org/projects/srallen086/intro2astro-hubble-testing/classify
A survey task: https://classifier-light-theme.pfe-preview.zooniverse.org/projects/srallen086/survey-test/classify

**Comment by eatyourgreens** [comment]
Looks good to me. Trying out a couple of question tasks, my only comment is that it's a little hard to see where the focus is when you tab to an answer that's already been selected.

**Comment by beckyrother** [comment]
Yep, let's talk about this. We can either slack call today, or I can be in the office tomorrow, whichever works for you.

Also wow I did not do a single "Done" button 🤦‍♀️ . Should those be green #078F52 rather than gold to indicate Done-ness? I like the blue for done & talk.

**Comment by srallen** [comment]
@beckyrother can we do a call tomorrow, maybe 9:30 or 10am?

**Comment by beckyrother** [comment]
@srallen that works! I'm free all morning.

File: package-lock.json [hunk 1]
```diff
@@ -7731,29 +8923,22 @@
       }
     },
     "react-select": {
-      "version": "1.0.0-rc.10",
-      "resolved": "https://registry.npmjs.org/react-select/-/react-select-1.0.0-rc.10.tgz",
-      "integrity": "sha1-8Tc0YlD5JVyXn7+iGGCJmSh3I1A=",
+      "version": "1.0.1",
```
Comment (eatyourgreens) [inline package-lock.json]: I think this changed the behaviour of the dropdown menus on Notes from Nature the last time we accidentally upgraded react-select. I thought the package file had been locked to a version of 1.0.0.

**Comment by srallen** [comment]
@eatyourgreens I'm not sure how to deal with merge conflicts in the package-lock.json other than trash it and regenerate it. In theory, minor updates should be safe according to semvar, but we know that the react-select project hasn't done semvar correctly. Checking the package.json, it is still using the tilde in its version specification. I can lock it after the classifier-dark-theme branch gets merged into this one?

**Comment by eatyourgreens** [comment]
@srallen I usually deal with these conflicts by [checking out package-lock.json from master](https://github.com/zooniverse/Panoptes-Front-End/blob/master/.github/CONTRIBUTING.md#merge-conflicts-with-package-lockjson), then running `npm install` to install any new packages. If that isn't working then I think the only thing you can do is trash it and reinstall everything.

**Comment by srallen** [comment]
@eatyourgreens Right, I think a remember now talking about this with you in the past. That's a much better idea. I think I can still do that. I'll clean up the file, if possible, before the final staging before merging into master.

**Comment by srallen** [comment]
@beckyrother would like to tentatively plan on announcing this early next week assuming this gets a final approval in the next few days? Do you want to send out a staging link to the UX test listserv or should we just merge to production and announce it via Talk?

@eatyourgreens or @mcbouslog do either of you have time to do a final check on this in the next few days?

**Comment by mcbouslog** [comment]
yes I'll review today or tomorrow.

**Comment by beckyrother** [comment]
@srallen I think since this has been such a requested issue and just finishes the theming we started, we don't need to send out a test first. I'll announce it over Talk when we merge it.

**Comment by eatyourgreens** [comment]
I can look at this tomorrow.

**Comment by mcbouslog** [comment]
Nothing blocking here, some points might be the intended design, or better left for a subsequent iteration. 

Went through a bunch of projects on production and made the following notes (which aren't unique to project(s) also noted, but where I happened to notice it):
- Back button smaller on second step ([Penguin Watch](https://classifier-light-theme.pfe-preview.zooniverse.org/projects/penguintom79/penguin-watch/classify?env=production), Manatee Chat, Galaxy Zoo), intended?
- dark theme - Done & Talk filled with blue, Back and Done transparent (Penguin Watch), should all these buttons be transparent, then fill on hover?
- dark theme - Back button doesn't have different color fill on hover, just border change, while other buttons have some color fill of button on hover (Penguin Watch)
- just Talk button text (not Done & Talk) not on same level as Next button text ([Galaxy Builder](https://classifier-light-theme.pfe-preview.zooniverse.org/projects/tingard/galaxy-builder/classify?env=production))
- ~~identified animal label covers video speed selection (Amazo'N'Oil)~~ unrelated to this PR
- "Going back..." warning remains even if not hover on Back button, disappears on hover ([Amazo'N'Oil](https://classifier-light-theme.pfe-preview.zooniverse.org/projects/marcartro/amazonoil/classify?env=production))
- dark theme, highlighted option before selection hard to see ([Galaxy Zoo](https://classifier-light-theme.pfe-preview.zooniverse.org/projects/zookeeper/galaxy-zoo/classify?env=production)), in light theme there's light fill of button, in dark theme border just changes subtly, not sure if want some button fill in dark theme too
- if no task help, Back and Next/Done very close to task - [staging project](https://classifier-light-theme.pfe-preview.zooniverse.org/projects/markb-panoptes/focus-testing-project/classify?reload=0&workflow=2955)

I can include screen shots to better illustrate any of the above. 
I'll go through code next, though I doubt I'll have much feedback on the code itself.

**Comment by eatyourgreens** [comment]
The Back button isn't disabled on the first task of a workflow, even though it can't be used. It also responds to hover on the first task, and displays the warning message.

**Comment by CKrawczyk** [comment]
If I go a few question deep into Galaxy Zoo and then use the back button to go back to the first question the warning message is toggled in the wrong direction (hovering over back makes the message go away instead of show up).  Also the warning message extends below the box making it harder to read (tested in Firefox 59.0.2).

screen shot 2018-04-27 at 1 41 53 pm

**Comment by eatyourgreens** [comment]
Line heights don't match for the task navigation buttons.
Screenshot of Done & Talk and Done, showing text at different heights.

**Comment by eatyourgreens** [comment]
Given how large this PR is, I'd be inclined to follow [the Springer playbook](https://github.com/springernature/frontend-playbook/blob/master/practices/code-review.md) and address each of these styling issues in their own separate PR.

**Comment by srallen** [comment]
I like @eatyourgreens's suggestion and I can branch off this and work on fixes separately. Are there any blocking issues? Would folks be ok with a merge on Monday even if the style issues are still present?

**Comment by eatyourgreens** [comment]
I'm okay with merging this and fixing style issues separately, rather than waiting on one massive PR.

**Comment by mcbouslog** [comment]
I also think merging this Monday then addressing style issues in separate PRs sounds good. I'll likely approve after a cursory review code today.

**Comment by eatyourgreens** [comment]
I went through a few projects and didn't find anything beyond the style issues already listed by @mcbouslog.

#4564 (the gold standard checkbox) is also broken on this branch. In my opinion, we shouldn't fix that on master but should wait until this PR is merged and fix it in the task components here.

**Comment by srallen** [comment]
Ok, thanks! This branch is getting unwieldily (sorry!). Once everyone is done with their reviews and comments and if there are no blocking issues, I'll collate the remaining issues and open up an issue with a list of to dos.

**Comment by eatyourgreens** [comment]
Not related to this PR as such, but [the How section](https://github.com/springernature/frontend-playbook/blob/master/practices/code-review.md#how) of that playbook is good advice for PR management.

File: app/classifier/task-nav.jsx [hunk 2]
```diff
@@ -122,78 +91,34 @@ class TaskNav extends React.Component {
...
+            areAnnotationsNotPersisted={!this.props.workflow.configuration.persist_annotations}
+            autoFocus={this.props.autoFocus}
+            classification={this.props.classification}
+            completeClassification={this.completeClassification}
+            completed={completed}
+            demoMode={this.props.demoMode}
+            destroyCurrentAnnotation={this.destroyCurrentAnnotation}
+            nextSubject={this.props.nextSubject}
+            project={this.props.project}
+            showBackButton={visibleTasks.length > 1 && !completed && (this.props.classification.annotations.indexOf(this.props.annotation) !== 0)}
```
Comment (mcbouslog) [inline app/classifier/task-nav.jsx]: I think `this.props.classification.annotations` should be `this.props.annotations`

Comment (srallen) [reply app/classifier/task-nav.jsx]: Good catch. I likely missed this from the PR that changed that prop.

Comment (eatyourgreens) [reply app/classifier/task-nav.jsx]: Should fix the Back button showing on the first task.

**Pull request merged**